File: blogeur_ia_python/blogueur_ia/media_storage.py
import os
import hashlib
import mimetypes
import httpx
from datetime import datetime

from blogueur_ia.config import settings

EXTENSIONS_AUTORISEES = {"image/jpeg": ".jpg", "image/png": ".png", "image/webp": ".webp"}


async def telecharger_et_heberger_image(url: str) -> str | None:
    if not url:
        return None
    taille_max = settings.storage_max_size_mb * 1024 * 1024
    try:
        async with httpx.AsyncClient(timeout=12, follow_redirects=True) as http:
            resp = await http.get(url, headers={"User-Agent": "Mozilla/5.0 (BlogueurIA/1.0)"})
            if resp.status_code != 200:
                return None
            content_type = resp.headers.get("content-type", "").split(";")[0].strip()
            ext = EXTENSIONS_AUTORISEES.get(content_type)
            if not ext:
                guessed = mimetypes.guess_type(url)[0]
                ext = EXTENSIONS_AUTORISEES.get(guessed)
                if not ext:
                    return None
            content = resp.content
            if len(content) > taille_max or len(content) < 500:
                return None

            digest = hashlib.sha256(content).hexdigest()[:24]
            sous_dossier = datetime.utcnow().strftime("%Y/%m")
            dossier_complet = os.path.join(settings.storage_dir, sous_dossier)
            os.makedirs(dossier_complet, exist_ok=True)

            nom_fichier = f"{digest}{ext}"
            chemin_local = os.path.join(dossier_complet, nom_fichier)
            if not os.path.exists(chemin_local):
                with open(chemin_local, "wb") as f:
                    f.write(content)

            return f"{settings.storage_base_url}/{sous_dossier}/{nom_fichier}"
    except Exception:
        return None

File: blogeur_ia_node/src/mediaStorage.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";
import axios from "axios";
import { settings } from "./config";

const EXTENSIONS_AUTORISEES: Record<string, string> = {
  "image/jpeg": ".jpg",
  "image/png": ".png",
  "image/webp": ".webp",
};

function extFromUrl(url: string): string | null {
  const match = url.split("?")[0].match(/\.(jpe?g|png|webp)$/i);
  if (!match) return null;
  const ext = match[1].toLowerCase();
  return ext === "jpeg" ? ".jpg" : `.${ext}`;
}

export async function telechargerEtHebergerImage(url: string): Promise<string | null> {
  if (!url) return null;
  const tailleMax = settings.storageMaxSizeMb * 1024 * 1024;
  try {
    const resp = await axios.get<ArrayBuffer>(url, {
      timeout: 12000,
      maxRedirects: 5,
      responseType: "arraybuffer",
      headers: { "User-Agent": "Mozilla/5.0 (BlogueurIA/1.0)" },
      validateStatus: () => true,
    });
    if (resp.status !== 200) return null;

    const contentType = String(resp.headers["content-type"] || "").split(";")[0].trim();
    let ext = EXTENSIONS_AUTORISEES[contentType];
    if (!ext) {
      const guessed = extFromUrl(url);
      if (!guessed) return null;
      ext = guessed;
    }

    const buffer = Buffer.from(resp.data);
    if (buffer.length > tailleMax || buffer.length < 500) return null;

    const digest = crypto.createHash("sha256").update(buffer).digest("hex").slice(0, 24);
    const now = new Date();
    const sousDossier = `${now.getUTCFullYear()}/${String(now.getUTCMonth() + 1).padStart(2, "0")}`;
    const dossierComplet = path.join(settings.storageDir, sousDossier);
    fs.mkdirSync(dossierComplet, { recursive: true });

    const nomFichier = `${digest}${ext}`;
    const cheminLocal = path.join(dossierComplet, nomFichier);
    if (!fs.existsSync(cheminLocal)) {
      fs.writeFileSync(cheminLocal, buffer);
    }

    return `${settings.storageBaseUrl}/${sousDossier}/${nomFichier}`;
  } catch {
    return null;
  }
}